### ISSUE TITLE
Default values unnecessarily passed to keywords when using dynamic library API

### ISSUE THREAD

**Issue by timmy8526** [robotframework/robotframework#2930]
let's assume we have a python customized keyword:
```
@keyword
def test_function(self, option_a, option_b='b', option_c=None):
     ....
  if not option_c:
      do something ...
```

Then we use it from Robotframework level like:
```
*** Test Cases ***
Test
    Test Keyword

*** Keywords ***
Test Keyword
    [Arguments]   ${a}   ${b}='b'
    test_function   option_a=${a}   option_b={b}
```

As you can see,
I didn't pass argument "option_c" from robotframework level, **and I expect option_c is still None.** 
However, my test keep failing, after debugging .. I surprisingly found that somehow the **the python default value **None** has been changed from <class 'NoneType'> to <class 'str'>** in Robotframework.

So the function will never go into IF statement as exception.. due to the **None** actually becomes **'None'** if function go thru Robotframework
 ``` 
  if not option_c:  --> expect option_c is None / but it becomes 'None'
      do something ...
```

Is this a expected behavior or a bug?

**Comment by pekkaklarck** [comment]
No, this shouldn't happen. Something close to this happens if you are using the dynamic library API, though. Can you check your example and also tell which library API you use and what is your Robot Framework version?

**Comment by timmy8526** [comment]
@pekkaklarck yes, I was using robotframework-seleniumlibrary **DynamicCore** class to fetch my customized keywords from python. 

Versions:
robotframework-seleniumlibrary==3.1.1
robotframework==3.0.2

It's seems like the issue is caused by below functions:
```
def get_keyword_arguments(self, name):
        kw = self.keywords[name] if name != '__init__' else self.__init__
        args, defaults, varargs, kwargs = self._get_arg_spec(kw)
        args += ['{}={}'.format(name, value) for name, value in defaults]  --> **this line will convert args name/value to string and save to array**
        if varargs:
            args.append('*{}'.format(varargs))
        if kwargs:
            args.append('**{}'.format(kwargs))
        return args
```

Originally I want to modify the code to find a workaround (e.g. using dict to save name/value instead of array) by myself, however there are too many functions involved from robotframework lib..

**Comment by pekkaklarck** [comment]
On mobile so only quick comments:
- This ought to happen only with the arguments in the middle if you pass arguments after them. It shouldn't thus happen with the original example.
- We are currently working with automatic type conversion and as a result also this problem is likely to be addressed in RF 3.1.
- At the moment the best workaround is implementing a similar utility like `is_noney` that SeleniumLibrary uses. Or using that directly if your library anyway works with SL.

**Comment by timmy8526** [comment]
@pekkaklarck 
Thanks for the advice! I've resolved it by modifying the code in run_keyword function.
```
 def run_keyword(self, name, args, kwargs):
        return self.keywords[name](*(self.get_real_args(args)), **kwargs)
```
So the get_real_args will handle the returned tuple and change 'None' to None as expectation.
Thanks again for your kind help! :)

**Comment by pekkaklarck** [comment]
Finally in front of a real keyboard and easier to explain this behavior in a bit more detail. There are two design decisions related to this:

1. Default values used with dynamic libraries are always communicated to Robot as strings in format `argument=default` which obviously loses type information. This generally doesn't matter much because this information is used for documentation purposes (Libdoc) and to tell to Robot that it's OK to omit certain arguments. For example, if a keyword has an argument spec `['arg1', 'arg2=default']`, Robot knows it's fine to call it like `Example    foo` and `Example    foo    bar`. In the former case arguments it's going to pass to the dynamic library are `['foo']` and in the latter they are `['foo', 'bar']`. In other words, the actual default value is normally not passed back to the library. There is, however, case where that happens explained.

2. Normal arguments passed to dynamic libraries are always passed as a list regardless how they are called. If a keyword has an argument spec `['a', 'b=foo', 'c=bar']` and it's called like `Example    xxx     b=yyy`, it is going to get arguments `['xxx', 'yyy']` regardless does the `run_keyword` method have a signature `name, args` or `name, args, kwargs`. An alternative in the latter case would be passing only `['xxx']` as args and  pass `{'b': 'yyy'}` as kwargs. The current design has a problem that if the example keyword is called like `Example    xxx    c=yyy` the provided default value needs to be used to be able to pass arguments as a list like `['xxx', 'foo', 'yyy']`. Passing `['xxx']` and `{'c': 'yyy'}` instead would solve the problem, but this change could cause backwards incompatibility problem and wouldn't naturally work if `run_keyword` has signature `name, args`. I've been thinking that this should be changed, but I'm not sure are benefits big enough.

**Comment by pekkaklarck** [comment]
One reason I think changing how arguments are passed to dynamic keywords is not worth the effort is that from usability point of view it would be best if the library handles explicitly passed string `None` similarly as if the value would be passed as a real Python `None`. For example, if you have a keyword with an argument spec `['arg1=None', 'arg2=None']` and it's called like `Example    None    foobar`, it would be good to consider that same as more verbose `Example    ${None}    foo`. If you do that, then the possible default value `None` is handled automatically as well. This is what SeleniumLibrary does.

**Comment by pekkaklarck** [comment]
As I mentioned earlier, we are currently working with automatic argument conversion. Issue #2890 covers doing that based on Python 3 type annotations, but I want to extend it to dynamic libraries as well and also to do it based on argument default values. Details are still open, but my expectation is that this problem will be mitigated by those enhancements.

**Comment by timmy8526** [comment]
>  The current design has a problem that if the example keyword is called like Example xxx c=yyy, the provided default value needs to be used to be able to pass arguments as a list like ['xxx', 'foo', 'yyy']. Passing ['xxx'] and {'c': 'yyy'} instead would solve the problem.

Exactly, originally I want to fix it by passing arguments like ['xxx'] and {'c': 'yyy'}, however as I mentioned earlier.. too many functions involved and I don't have enough knowledge of the whole argument passing flow in RF, so I choose another quick fix from run_keyword level. 

However, I still look forward to RF can handle this type of issue eventually. 
Thank you for your hard work 👍

**Comment by pekkaklarck** [comment]
Changing the dynamic API to pass all arguments used like `name=value` in kwargs (when `run_keywords` accepts it) ought to be pretty easy (when you know the code base related to this) but I'm worried about possible backwards incompatibility issues. I guess now during RF 3.1 development would be a good time to test that. The change could still be easily reverted if users testing the preview releases encounter problems. I'll ask opinion from others on the #devel channel on Slack. Are you @timmy8526 there?

**Comment by timmy8526** [comment]
@pekkaklarck No, I am not. I actually just start to learn RF since June, not sure if I am useful enough to be in the channel lol.

**Comment by pekkaklarck** [comment]
Based on our discussion here you are definitely useful! You can join our Slack community here:
https://robotframework-slack-invite.herokuapp.com/

Yesterday on the #devel channel we agreed to test would changing the logic to pass arguments used like `name=value` always in kwargs break anything. I'm fairly certain it's safe with Python based libraries, but I'm somewhat worried there could be problems with JavaLibCore and SwingLibrary that use it. Luckily @mihaiparvu promised to test that once we've changed the code. That won't happen in the immediate future, though, because getting the type conversion functionality implemented and the next 3.1 preview release out is more important. To get feedback early, this needs to be in the next preview release, though.

**Comment by pekkaklarck** [comment]
As 9ec5a02ac4db6f419a95e14206854268ba3423ee shows, changing functionality was trivial. Biggest task was actually cleaning up related tests but that was done separately in ed1acf182ab070394e2fa818f38e53badf40673a and a6e1785f2b2d1b26713b18ddb32c4e27e556f467.

I don't expect there to be real problems with Python based libraries, because with them the natural way to forward arguments to actual keywords is using style `kw(*args, **kwargs)` which only works better after this change. I don't know how dynamic libraries implemented in Java generally handle forwarding arguments, but also they are only affected if they actually accept `kwargs`. If there are problems, we can still change this so that this change doesn't have any effect with Java based libraries. @mihaiparvu promised to test this with SwingLibrary and hopefully also others can test this latest when RF 3.1 beta 1 is out.
